File: typescript/robot/index.test.ts
import Robot from '.';
import type { RobotParams } from '../@types';

export const worldDimensions = {
  x: 5,
  y: 3,
};

export const robotOneParams: RobotParams = {
  instructions: ['r', 'f', 'r', 'f', 'r', 'f', 'r', 'f'],
  initialPosition: {
    x: 1,
    y: 1,
    orientation: 'e',
  },
};

export const robotTwoParams: RobotParams = {
  instructions: [
    'f',
    'r',
    'r',
    'f',
    'l',
    'l',
    'f',
    'f',
    'r',
    'r',
    'f',
    'l',
    'l',
  ],
  initialPosition: {
    x: 3,
    y: 2,
    orientation: 'n',
  },
};

export const robotThreeParams: RobotParams = {
  instructions: ['l', 'l', 'f', 'f', 'f', 'l', 'f', 'l', 'f', 'l'],
  initialPosition: {
    x: 0,
    y: 3,
    orientation: 'w',
  },
};

describe('Robot', () => {
  it('accept params', () => {
    const {
      worldDimensions: dimensions,
      instructions,
      initialPosition,
    } = robotOneParams;

    const robot = new Robot(robotOneParams);

    expect(robot.data).toEqual({
      dimensions,
      instructions,
      position: initialPosition,
      isLost: false,
    });
  });

  it('moves to a new position', () => {
    const robotOne = new Robot(robotOneParams);

    robotOne.setworldDimensions(worldDimensions);
    robotOne.move();

    const robotTwo = new Robot(robotTwoParams);

    robotTwo.setworldDimensions(worldDimensions);
    robotTwo.move();

    expect(robotOne.data.position).toEqual({
      x: 1,
      y: 1,
      orientation: 'e',
    });

    expect(robotTwo.data.position).toEqual({
      x: 3,
      y: 4,
      orientation: 'n',
    });
  });

  it('reports last known coordinates when is lost', () => {
    const robotOne = new Robot(robotOneParams);

    robotOne.setworldDimensions(worldDimensions);
    robotOne.move();

    const robotTwo = new Robot(robotTwoParams);

    robotTwo.setworldDimensions(worldDimensions);
    robotTwo.move();

    expect(robotOne.data.position).toEqual({
      x: 1,
      y: 1,
      orientation: 'e',
    });

    expect(robotOne.data.isLost).toBe(false);

    expect(robotTwo.lostCoordinate).toEqual({
      x: 3,
      y: 3,
      orientation: 'n',
    });

    expect(robotTwo.data.isLost).toBe(true);
  });

  it("don't fell off of the grid when is in an scented coordinate are set by constructor", () => {
    const params: RobotParams = {
      ...robotThreeParams,
      lostRobotCoordinates: {
        '3': {
          '3': 'n',
        },
      },
    };

    const robotThree = new Robot(params);

    robotThree.setworldDimensions(worldDimensions);
    robotThree.move();

    expect(robotThree.data.position).toEqual({
      x: 2,
      y: 3,
      orientation: 's',
    });

    expect(robotThree.data.isLost).toBe(false);
    expect(robotThree.lostCoordinate).toBeUndefined();
  });

  it("don't fell off of the grid when is in an scented coordinate are set by setter", () => {
    const robotThree = new Robot(robotThreeParams);

    robotThree.setLostRobotCoordinates({
      '3': {
        '3': 'n',
      },
    });

    robotThree.setworldDimensions(worldDimensions);
    robotThree.move();

    expect(robotThree.data.position).toEqual({
      x: 2,
      y: 3,
      orientation: 's',
    });

    expect(robotThree.data.isLost).toBe(false);
    expect(robotThree.lostCoordinate).toBeUndefined();
  });
});

File: python/robot/index_test.py
# import pytest
from index import Robot

world_dimensions = {
    "x": 5,
    "y": 3,
}

robot_one_params = {
    "instructions": ["r", "f", "r", "f", "r", "f", "r", "f"],
    "position": {
        "x": 1,
        "y": 1,
        "orientation": "e",
    },
}


robot_two_params = {
    "instructions": ["f", "r", "r", "f", "l", "l", "f", "f", "r", "r", "f", "l", "l"],
    "position": {
        "x": 3,
        "y": 2,
        "orientation": "n",
    },
}

robot_three_params = {
    "instructions": ["l", "l", "f", "f", "f", "l", "f", "l", "f", "l"],
    "position": {
        "x": 0,
        "y": 3,
        "orientation": "w",
    },
}


def test_accept_params():
    robot_one = Robot(
        instructions=robot_one_params["instructions"],
        position=robot_one_params["position"],
    )
    assert (robot_one.instructions) == ["r", "f", "r", "f", "r", "f", "r", "f"]


def test_move_to_a_new_position():
    robot_one = Robot(
        instructions=robot_one_params["instructions"],
        position=robot_one_params["position"],
    )
    robot_one.set_world_dimmensions(world_dimensions)
    robot_one.move()

    robot_two = Robot(
        instructions=robot_two_params["instructions"],
        position=robot_two_params["position"],
    )
    robot_two.set_world_dimmensions(world_dimensions)
    robot_two.move()

    assert robot_one.position == {"x": 1, "y": 1, "orientation": "e"}
    assert robot_two.position == {"x": 3, "y": 4, "orientation": "n"}


def test_reports_last_known_coordinates_when_is_lost():
    robot_two_params = {
        "instructions": [
            "f",
            "r",
            "r",
            "f",
            "l",
            "l",
            "f",
            "f",
            "r",
            "r",
            "f",
            "l",
            "l",
        ],
        "position": {
            "x": 3,
            "y": 2,
            "orientation": "n",
        },
    }

    robot = Robot(
        instructions=robot_two_params["instructions"],
        position=robot_two_params["position"],
    )

    robot.set_world_dimmensions(world_dimensions)
    robot.move()

    assert robot.position == {"x": 3, "y": 4, "orientation": "n"}
    assert robot.is_lost == True
    assert robot.lost_coordinate == {"x": 3, "y": 3, "orientation": "n"}


def test_prevent_get_lost_if_robot_is_in_a_lost_coordinate():
    robot = Robot(
        instructions=robot_three_params["instructions"],
        position=robot_three_params["position"],
    )

    robot.set_world_dimmensions(world_dimensions)
    robot.set_lost_robot_coordinates({3: {3: "n"}})
    robot.move()

    assert robot.position == {"x": 2, "y": 3, "orientation": "s"}
    assert robot.is_lost == False
